// animelyrics/index.js
var request = require('request-promise-native');
const cheerio = require('cheerio');

request = request.defaults({jar: true});

const options = {
    uri: 'http://www.google.com/ncr',
    headers: {
        'User-Agent': 'Mozilla/5.0 (Macintosh; U; Intel Mac OS X 10.6; rv:1.9.2.16) Gecko/20110319 Firefox/3.6.16'
    }
};

module.exports = async function search(query){
    let song = {
        anime: '',
        link: '',
        name: '',
        description: '',
        info: [],
        lyrics: '',
        english: null
    }

    try{

        options['uri'] = `https://www.google.com/search?gws_rd=ssl&site=&source=hp&q=anime%20lyrics%20dot%20com%20${query.replace(/\s/g, '%20')}&oq=google`;
        let body = await request(options)

        let $ = cheerio.load(body);

        let htm = 'https://' + $('cite').first().text();
        let link = htm.replace('.htm', '.txt');

        song.link = link;

        options['uri'] = link;
        options.headers['Referer'] = htm;

        let b = await request(options);
        $ = cheerio.load(b);

        let sections = $('body').text().split('-'.repeat(107));
        
        let d = sections[0]

        song.anime = d.match(/.* - .* - .*/)[0].split(' - ')[1]
        song.description = d.match(/Description: .*/g)[1].split(': ')[1]
        for(let inf of d.match(/.*: .*/g)){
            if (!inf.includes('Description')){
                song.info.push(inf);
            }
        }

        if(sections[0].includes('Translation')){
            let english = [], romaji = [];
            for(let line of sections[1].split('\n')){
                if (line.startsWith('\t') && line != '\t\t'){
                    english.push(line.replace(/\t/g, ''));
                }else{
                    romaji.push(line);
                }
            }
            song.english = english;
            song.lyrics = romaji;
        }else{
            let romaji = sections[1].split('\n')
            song.lyrics = romaji;
        }

        return song

    }catch(err){
        throw err;
    }
}



/**
        options['uri'] = link;

        let b = await request(options);
        $ = cheerio.load(b);

        console.log($('html').html());

        let text = $('td').first().text();

        song.name = $('h1:not(#logo)').first().text();
        song.anime = $(`a[href='anime/${link.substring(8).split('/')[2]}/']`).text();

        let descIndex = text.indexOf('Description: ');

        if(text.includes('English Translation')){
            let lyrics = english = '';
            $(".romaji").get().forEach((el, i, arr) => {
                lyrics += $(el).text().replace('Lyrics from Animelyrics.com', '');
            });
            $(".translation").get().forEach((el, i, arr) => {
                english += $(el).text().replace('Lyrics from Animelyrics.com', '');
            });
            song.lyrics = lyrics;
            song.english = english;

            let songInfo = text.substring(descIndex, text.indexOf('View Kanji')).replace(/(\t|\s{2,})/g, '\n');
            song.description = songInfo.substring(13, songInfo.indexOf('\n'));
            song.info = songInfo.substring(songInfo.indexOf('\n') + 1);
        }
        else
        {
            let lyricIndex = text.indexOf('Lyrics from Animelyrics.com') + 27;
            song.lyrics = text.substring(lyricIndex, text.indexOf('Transliterated')).replace(/(\t|\s{2,})/g, '');

            let pre = text.substring(descIndex, lyricIndex).replace(/(\t|\s{2,})/g, '\n').split('\n').slice(0, -3);
            song.description = pre[0].substring(12);
            song.info = pre[1];
        }

        return song;

        ===============================================================
            request(link, (err, res, b) => {
                $ = cheerio.load(b);

                let text = $('td').first().text();

                song.name = $('h1:not(#logo)').first().text();

                let lyricIndex = text.indexOf('Lyrics from Animelyrics.com') + 27;
                song.lyrics = text.substring(lyricIndex);

                //text = text.replace(/[\s+][\s+]/g,'');
                let descIndex = text.indexOf('Description: ');

                let pre = text.substring(descIndex, lyricIndex).replace(/\s{2,}/g, '\n').split('\n').slice(0, -3);

                console.log(pre);
                return ;
                //console.log(song);
            })
        });
    });
    **/
